File: Homework/Ngan_Luu/238-product-of-array-except-self.py
https://leetcode.com/problems/product-of-array-except-self/
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        output=[]
        prefixProduct=[nums[0]]*n
        suffixProduct=[nums[-1]]*n
        for i in range(1,n):
            prefixProduct[i]=prefixProduct[i-1]*nums[i]
            suffixProduct[n-1-i]=suffixProduct[n-i]*nums[n-1-i]
        output.append(suffixProduct[1])  #i=0
        for i in range(1,n-1):
            output.append(prefixProduct[i-1]*suffixProduct[i+1])
        output.append(prefixProduct[n-2]) #i=n-1
        return output

File: Homework/Son/W2/238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
 var productExceptSelf = function(nums) {
    let left = 1;
    let right = 1; 
    let answer = []
    for (let i = 0; i <nums.length; i++) {
        if (i > 0) {left = left*nums[i-1]}
        answer.push(left)
    }
    for (let i = nums.length -1; i >= 0; i--) {
        if (i < nums.length - 1 ) {right = right*nums[i + 1]}
        answer[i] = answer[i]*right
    }
    return answer
};

// https://leetcode.com/problems/product-of-array-except-self/